File: 0093-Restore IP Addresses/0093-Restore IP Addresses.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> result;
        string path;
        dfs(s, 0, 0, path, result);
        return result;
    }
    
private:
    void dfs(string& s, int start, int count, string path, vector<string>& result) {
        if (count == 4) {
            if (start == s.size()) {
                result.push_back(path);
            }
            return;
        }
        
        if (path != "") {
            path += '.';
        }
        
        for (int i = start, j = 1; i < s.size() && j <= 3; ++i, ++j) {
            string temp = s.substr(start, i - start + 1);
            if (stoi(temp) >= 0 && stoi(temp) <= 255 && (temp.size() == 1 || temp[0] != '0')) {
                dfs(s, i + 1, count + 1, path + temp, result);
            }
        }
    }
};

File: 0001-0100/0093-Restore IP Addresses/0093-Restore IP Addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        result = []
        path = []
        def dfs(start):
            if len(path) == 4:
                if start == len(s):
                    result.append('.'.join(path))
                return
            
            for i in range(start, min(start + 3, len(s))):
                temp = s[start: i+1]
                num = int(temp)
                if 0 <= num <= 255 and (len(temp) == 1 or temp[0] != '0'):
                    path.append(temp)
                    dfs(i+1)
                    path.pop()
        dfs(0)
        return result
